core/modules/wiki.js

```diff
@@ -558,39 +558,49 @@ exports.sortByList = function(array,listTitle) {
 			}
 		}
 		// Finally obey the list-before and list-after fields of each tiddler in turn
-		var sortedTitles = titles.slice(0);
-		for(t=0; t<sortedTitles.length; t++) {
-			title = sortedTitles[t];
-			var currPos = titles.indexOf(title),
-				newPos = -1,
-				tiddler = this.getTiddler(title);
-			if(tiddler) {
-				var beforeTitle = tiddler.fields["list-before"],
-					afterTitle = tiddler.fields["list-after"];
-				if(beforeTitle === "") {
-					newPos = 0;
-				} else if(afterTitle === "") {
-					newPos = titles.length;
-				} else if(beforeTitle) {
-					newPos = titles.indexOf(beforeTitle);
-				} else if(afterTitle) {
-					newPos = titles.indexOf(afterTitle);
-					if(newPos >= 0) {
-						++newPos;
+		var sortedTitles = titles.slice(0),
+			replacedTitles = Object.create(null),
+			self = this;
+		function replaceItem(title) {
+			if(!$tw.utils.hop(replacedTitles, title)) {
+				replacedTitles[title] = true;
+				var newPos = -1,
```

Comment:
The upper two lines guard against infinite loops.

Reply:
That's ok. .. BUT where is the difference now, between your algorithm and the existing one. ... 

If the users expectes a different sorting result as the algorithm can create, the user still needs to manually fix it. ... 

So why shouldn't we tell them in the first place, that they have to manually fix the relations.

Reply:
Circular dependencies still produce undefined behavior.

My fix aims to correct issues that may occur with chained dependencies.
```
A (list-after B)
B (list-after C) 
```
This simple case _does_ have defined behavior, but because the old algorithm always processed them in order, it would mess that example up. A would not be after B. That's what recursive calls fix.